shelljs/shelljs.d.ts

```diff
@@ -434,31 +436,44 @@ declare module "shelljs"
      * Object containing environment variables (both getter and setter). Shortcut to process.env.
      */
     export var env: { [key: string]: string };
-
-    /*
-
-    // Not yet implemented due to implementation issues (constant overloads and return types).
-    // See: https://github.com/arturadib/shelljs#execcommand--options--callback
-
-    export function exec(command: string, options: ExecOptions, callback: (code: number, output: string) => any): any;
-    export function exec(command: string, options: ExecOptions): any;
-
-    interface ExecOptions
-    {
-        silent: boolean;
-        async: boolean;
-    }
-
-    */
-
+    
     /**
      * Executes the given command synchronously.
-     * @param  {string}          command The commadn to execute.
+     * @param  {string}          command The command to execute.
      * @return {ExecReturnValue}         Returns an object containing the return code and output as string.
      */
-    export function exec(command: string): ExecReturnValue;
+    export function exec(command: string): ExecOutputReturnValue;
+    /**
+     * Executes the given command synchronously.
+     * @param  {string}          command The command to execute.
+     * @param  {ExecOptions}     options Silence and synchronous options.
+     * @return {ExecReturnValue}         Returns an object containing the return code and output as string.
+     */
+    export function exec(command: string, options: ExecOptions): ExecOutputReturnValue;
```

Comment:
OOOOOOh, okay, I wasn't following you before.  Your suggestion would make sense if either ChildProcess or ExecOutputReturnValue were classes and could be used with instanceof, like:

```javascript
var version = shell.exec("node --version", { async: false });
if (version instanceof child.ChildProcess) {
    // ...
}
```
but as it stands, they're both interfaces, so the only way to disambiguate the union type is to use type-casting (e.g. <ChildProcess>).  Forcing this on the user leads to cumbersome syntax for almost zero type-safety gain, so i would suggest we just make this one overload return `any`, e.g.:

```javascript

    export function exec(command: string): ExecOutputReturnValue;
    export function exec(command: string, options: ExecOptions): any;
    export function exec(command: string, options: ExecOptions, callback: ExecCallback): child.ChildProcess;
    export function exec(command: string, callback: ExecCallback): child.ChildProcess;
```
```
```